libsolidity/ast/ASTUtils.cpp

```diff
@@ -18,12 +18,47 @@
 
 #include <libsolidity/ast/AST.h>
 #include <libsolidity/ast/ASTUtils.h>
+#include <libsolidity/ast/ASTVisitor.h>
 
 #include <libsolutil/Algorithms.h>
 
 namespace solidity::frontend
 {
 
+namespace
+{
+
+class ASTNodeLocator: public ASTConstVisitor
+{
+public:
+	explicit ASTNodeLocator(int _pos): m_offsetInFile{_pos}, m_innermostMatch{nullptr} {}
+
+	bool visitNode(ASTNode const& _node) override
+	{
+		// In the AST parent location always covers the whole child location.
+		// The parent is visited first so to get the innermost node we simply
+		// take the last one that still contains the offset.
+
+		if (!_node.location().containsOffset(m_offsetInFile))
+			return false;
+
+		m_innermostMatch = &_node;
+		return true;
+	}
+
```

Comment:
private

Reply:
I know it has to be read by `locateInnermostASTNode`, but I think mutable members should only be public if it is OK to modify them at any time.